charm/src/charm.py

```diff
@@ -0,0 +1,186 @@
+#!/usr/bin/env python3
+# Copyright 2023 Paul Larson
+# See LICENSE file for licensing details.
+
+import logging
+import ops
+
+from ops.pebble import Layer
+from charms.data_platform_libs.v0.data_interfaces import DatabaseCreatedEvent
+from charms.data_platform_libs.v0.data_interfaces import DatabaseRequires
+
+logger = logging.getLogger(__name__)
+
+
+class TestflingerCharm(ops.CharmBase):
+    """Testflinger charm"""
+
+    _stored = ops.framework.StoredState()
+
+    def __init__(self, *args):
+        """Initialize the charm"""
+        super().__init__(*args)
+        self.pebble_service_name = "testflinger"
+        self.container = self.unit.get_container("testflinger")
+        self._stored.set_default(
+            reldata={},
+        )
+
+        self.mongodb = DatabaseRequires(
+            self,
+            relation_name="mongodb_client",
+            database_name="testflinger_db",
+        )
+
+        self.framework.observe(
+            self.mongodb.on.database_created,
+            self._on_mongodb_client_relation_changed,
+        )
+        self.framework.observe(
+            self.mongodb.on.endpoints_changed,
+            self._on_mongodb_client_relation_changed,
+        )
+        self.framework.observe(
+            self.on.mongodb_client_relation_broken,
+            self._on_mongodb_client_relation_removed,
+        )
+        self.framework.observe(
+            self.on.testflinger_pebble_ready, self._on_testflinger_pebble_ready
+        )
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+
+    @property
+    def version(self) -> str:
+        """Report the current version of the app"""
+        # TODO: get the version somehow and return it
+        return "Version ?"
```

Comment:
What if we get it from a docker image version? Is it possible?

Reply:
Not from what I can tell. We should probably add some sort of get version api in testflinger itself. It's not critical but I liked this idea from the examples I saw and wanted to leave it there as a placeholder in the charm.